src/snowflake/snowpark/_internal/analyzer/analyzer.py

```diff
@@ -688,7 +691,20 @@ def binary_operator_extractor(
                 expr.right, df_aliased_col_name_to_real_col_name, parse_local_name
             )
         if isinstance(expr, BinaryArithmeticExpression):
+            # TODO: it doesn't seem appropriate to rewrite the expression at this stage,
```

Comment:
You're probably already on it, but I feel the logic here should be something like:
```
if is_user_defined_type(expr.left.datatype) or is_user_defined_type(expr.right.datatype):
   gen = get_user_defined_binary_expr(expr) # <-- look up based on left/right datatype and type(expr)
   if not gen:
      raise NotImplementedError(f"No sql codegen available for {type(expr)}: {expr.left.datatype} x {expr.right.datatype}"
```

and then you can register code generators like this:

```
register_sql_binary_codegen(Subtract, TimestampType, TimestampType, lambda left_sql, right_sql: f'datediff("ns", {right_sql_expr}, {left_sql_expr})')
....
```

Reply:
Done!